ledger/apply/application.go

```diff
@@ -181,28 +187,85 @@ func deleteApplication(balances Balances, creator basics.Address, appIdx basics.
 	// Updating the application. Fetch the creator's balance record
 	params, _, err := balances.GetAppParams(creator, appIdx)
 	if err != nil {
 		return err
 	}
 
 	proto := balances.ConsensusParams()
-
-	// The pre-application well-formedness check rejects big programs
-	// conservatively, it doesn't know the actual params.ExtraProgramPages, so
-	// it allows any programs that fit under the absolute max.
-	if err = ac.WellSizedPrograms(params.ExtraProgramPages, proto); err != nil {
-		return err
+	sizeChange := ac.UpdatingSizes()
+
+	if !sizeChange {
+		// The wellFormed() check rejects big programs conservatively, but it
+		// doesn't know the actual params.ExtraProgramPages, so it allows any
+		// programs that fit under the absolute max. (if there is a size change,
+		// that check is precise because the programs are in the transaction)
+		if err = ac.WellSizedPrograms(params.ExtraProgramPages, proto); err != nil {
+			return err
+		}
 	}
-
 	params.ApprovalProgram = ac.ApprovalProgram
 	params.ClearStateProgram = ac.ClearStateProgram
 	if proto.EnableAppVersioning {
 		params.Version++
 	}
 
+	// Install the new epp and schema (if its sufficient for current globals)
+	if sizeChange {
+		// We'll call the account that is currently on the hook for MBR space
+		// the "size sponsor".  It begins as the creator, but changes whenever
+		// there is a sizeChange update.
+
+		sponsor := params.SizeSponsor
+		if sponsor.IsZero() {
+			sponsor = creator
+		}
+
+		// Since the sponsor and the updater may be the same account, we make the
+		// entire change to the sponsor, including Put(), before we Get() the
+		// updater. (similar to how Move() works)
+
+		sponsorRecord, err := balances.Get(sponsor, false)
+		if err != nil {
+			return err
+		}
+		sponsorRecord.TotalAppSchema =
+			sponsorRecord.TotalAppSchema.SubSchema(params.GlobalStateSchema)
+		sponsorRecord.TotalExtraAppPages =
+			basics.SubSaturate(sponsorRecord.TotalExtraAppPages, params.ExtraProgramPages)
+		err = balances.Put(sponsor, sponsorRecord)
+		if err != nil {
+			return err
+		}
+
+		err = balances.SetAppGlobalSchema(creator, appIdx, ac.GlobalStateSchema)
+		if err != nil {
+			return fmt.Errorf("unable to change global schema: %w", err)
+		}
+		params.GlobalStateSchema = ac.GlobalStateSchema
+		params.ExtraProgramPages = ac.ExtraProgramPages
+		if updater == creator {
+			params.SizeSponsor = basics.Address{}
+		} else {
+			params.SizeSponsor = updater
+		}
+
+		updaterRecord, err := balances.Get(updater, false)
+		if err != nil {
+			return err
+		}
+		updaterRecord.TotalAppSchema =
+			updaterRecord.TotalAppSchema.AddSchema(params.GlobalStateSchema)
+		updaterRecord.TotalExtraAppPages =
+			basics.AddSaturate(updaterRecord.TotalExtraAppPages, params.ExtraProgramPages)
+		err = balances.Put(updater, updaterRecord)
+		if err != nil {
+			return err
+		}
+	}
+
 	return balances.PutAppParams(creator, appIdx, params)
```

Comment:
So you could have waited until here to update the limits instead of calling `SetAppGlobalSchema` and had the limits checked and set in `PutAppParams` instead right? or would that have been too late because you wanted to abort earlier?